### PR TITLE
Adding support for universal tag

### LINKED ISSUE #1083: support universal tag api
https://aws.amazon.com/blogs/aws/new-aws-resource-tagging-api/

### PR DESCRIPTION
**Pull request by alfredgamulo**
Add universal tag to route53 hostedzone and healthcheck
Add universal tag to glacier
Add tag augment to kms. Found that KMS while advertised as a supported service by universal tag, not actually supported.
resolves #1083

### CODE REVIEW

File: c7n/resources/glacier.py [hunk 1]
```diff
@@ -13,23 +13,54 @@
...
         service = 'glacier'
         enum_spec = ('list_vaults', 'VaultList', None)
         name = "VaultName"
         id = "VaultARN"
         dimension = None
 
-    resource_type = Meta
+    def augment(self, resources):
+        def process_tags(resource):
+            client = local_session(self.session_factory).client('glacier')
```
Comment (kapilt) [inline c7n/resources/glacier.py]: why is this using a resource specific tag api instead of the resourcegrouptag api?

File: c7n/resources/kms.py [hunk 2]
```diff
@@ -23,13 +23,21 @@
...
+
     def augment(self, resources):
         client = local_session(
             self.session_factory).client('kms')
         for r in resources:
             key_id = r.get('AliasArn') or r.get('KeyArn')
             info = client.describe_key(KeyId=key_id)['KeyMetadata']
             r.update(info)
+
+            tags = client.list_resource_tags(KeyId=key_id)['Tags']
```
Comment (kapilt) [inline c7n/resources/kms.py]: same question here, wrt to why a resource specific tag api instead of the resource group tag api.

File: c7n/tags.py [hunk 3]
```diff
@@ -671,3 +660,154 @@ def process(self, resources):
...
+        for r in resource_set:
+            _id = r[self.id_key]
+            if 'arn' in _id[:3]:
+                arns.append(_id)
+            else:
+                # catch edge-case e.g. ID is like /hostedzone/ZXB8XVP18AKH8
+                _id = _id.split("/")[-1]
+                arns.append(self.manager.generate_arn(_id))
+
+        self.manager.retry(
```
Comment (kapilt) [inline c7n/tags.py]: not sure if manager.retry is universal

File: c7n/tags.py [hunk 4]
```diff
@@ -671,3 +660,154 @@ def process(self, resources):
...
+            self.process_resource_set, self.id_key, resources, tags, self.log)
+
+    def process_resource_set(self, resource_set, tags):
+        client = utils.local_session(
+            self.manager.session_factory).client('resourcegroupstaggingapi')
+
+        arns = []
+        for r in resource_set:
+            _id = r[self.id_key]
+            if 'arn' in _id[:3]:
```
Comment (kapilt) [inline c7n/tags.py]: table for now, but should probably get an arn class going rather than adhoc arn parsing.

File: c7n/tags.py [hunk 5]
```diff
@@ -671,3 +660,154 @@ def process(self, resources):
...
+                arns.append(_id)
+            else:
+                # catch edge-case e.g. ID is like /hostedzone/ZXB8XVP18AKH8
+                _id = _id.split("/")[-1]
+                arns.append(self.manager.generate_arn(_id))
```
Comment (kapilt) [inline c7n/tags.py]: also unclear if generate_arn is universally implemented if we're going to depend on it, we probably need a unit test that validates presence on resource managers or a generic implementation.

File: tests/data/placebo/test_glacier_untag/tagging.UntagResources_1.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+            "HTTPStatusCode": 200, 
+            "RequestId": "488bc109-36ad-11e7-bfbf-47f7816c1ac2", 
+            "HTTPHeaders": {
+                "x-amzn-requestid": "488bc109-36ad-11e7-bfbf-47f7816c1ac2", 
+                "date": "Fri, 12 May 2017 00:52:33 GMT", 
+                "content-length": "25", 
+                "content-type": "application/x-amz-json-1.1"
+            }
+        }, 
+        "FailedResourcesMap": {}
```
Comment (kapilt) [inline tests/data/placebo/test_glacier_untag/tagging.UntagResources_1.json]: note this notion of inline failure in the response instead of an exception raised on error, we need to explicitly check this and log.

File: c7n/resources/glacier.py (hunk 1 above)
Comment (alfredgamulo) [inline c7n/resources/glacier.py]: resourcegrouptag api doesn't support fetching tags given a resource id.

File: c7n/tags.py (hunk 5 above)
Comment (alfredgamulo) [inline c7n/tags.py]: generate_arn is the naming convention that I found in other resources that needed the support of creating the ARN on the fly when it doesn't come back in the initial list call. Since ARN formats vary, my thought process here was that generate_arn was a necessary method to be added on a per-resource basis just as the augment is necessary for every resource that supports tag.

File: c7n/resources/glacier.py (hunk 1 above)
Comment (kapilt) [inline c7n/resources/glacier.py]: true, but it does support our common usage mode for a resource manager, which is fetching tags for all resources of a given type.

Comment (alfredgamulo) [inline c7n/resources/glacier.py]: Got it. I understand now, will fix it soon.

Comment (kapilt) [inline c7n/resources/glacier.py]: no worries, we can do that as a separate pull request, let me if you have a chance to sync up on this pr.

File: tests/test_glacier.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        client.create_vault(vaultName=name)
+        self.addCleanup(client.delete_vault, vaultName=name)
+
+        p = self.load_policy({
+            'name': 'glacier',
+            'resource': 'glacier',
+            'filters': [
+                {
+                    'type': 'value',
+                    'key': 'VaultName',
```
Comment (kapilt) [inline tests/test_glacier.py]: this policy and assertion doesn't seem to be doing anything. afaics it can just be omitted.

File: tests/test_glacier.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
+from common import BaseTest
+from c7n.executor import MainThreadExecutor
+from c7n.resources.glacier import Glacier
+
+
+class GlacierTagTest(BaseTest):
+
+    def test_glacier_tag(self):
```
Comment (kapilt) [inline tests/test_glacier.py]: given this is self contained it can use the functional decorator in tests/common.py

File: tests/test_glacier.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+        tags = client.list_tags_for_vault(vaultName=resources[0]['VaultName'])
+        self.assertEqual(len(tags['Tags']), 1)
+        self.assertTrue('abc' in tags['Tags'])
+
+    def test_glacier_untag(self):
+        session_factory = self.replay_flight_data('test_glacier_untag')
+        client = session_factory().client('glacier')
+
+        p = self.load_policy({
```
Comment (kapilt) [inline tests/test_glacier.py]: strong preference for using the api instead of re-running policies for test verification, it leads to smaller and faster tests.

File: tests/test_kms.py [hunk 10]
```diff
@@ -39,4 +39,36 @@ def test_key_rotation(self):
             session_factory=session_factory)
 
         resources = p.run()
-        self.assertEqual(len(resources), 7)
+        self.assertEqual(len(resources), 2)
+
+    def test_kms(self):
```
Comment (kapilt) [inline tests/test_kms.py]: what is this test asserting?.. afaics its just doing standard value semantics on kms keys, its unclear that its adding value as a result. if it is please add a comment to what its actually testing or make that more explicit.

Comment (alfredgamulo) [inline tests/test_kms.py]: I think my motivation was to ensure that kms key-alias augment ensures mapping with a valid key. (https://github.com/capitalone/cloud-custodian/blob/master/c7n/resources/kms.py#L48) 

You're right this isn't really adding value as a test. I'll review and remove this and any other test that doesn't make sense.
